File: apps/auth-server/src/main/java/com/example/chat/auth/server/core/domain/MfaRequirement.java
package com.example.chat.auth.server.core.domain;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

/**
 * MFA 요구사항
 */
public record MfaRequirement(
        boolean required,
        Set<MfaType> requiredMethods,
        Set<MfaType> completedMethods,
        String sessionId) {
    /** 모든 필수 MFA가 완료되었는가? */
    public boolean isComplete() {
        return completedMethods.containsAll(requiredMethods);
    }

    /** 남은 MFA 방식 */
    public Set<MfaType> getRemainingMethods() {
        Set<MfaType> remaining = new HashSet<>(requiredMethods);
        remaining.removeAll(completedMethods);
        return Collections.unmodifiableSet(remaining);
    }

    public static MfaRequirement none(String sessionId) {
        return new MfaRequirement(false, Collections.emptySet(), Collections.emptySet(), sessionId);
    }

    public static MfaRequirement otp(String sessionId) {
        return new MfaRequirement(true, Collections.singleton(MfaType.OTP), Collections.emptySet(), sessionId);
    }

    public static MfaRequirement withCompleted(
            String sessionId,
            Set<MfaType> required,
            Set<MfaType> completed) {
        return new MfaRequirement(true, required, completed, sessionId);
    }
}

File: apps/auth-server/src/main/kotlin/com/example/chat/auth/server/core/domain/MfaRequirement.kt
package com.example.chat.auth.server.core.domain

/**
 * MFA 요구사항
 * - "MFA가 필요한가?"를 판단하는 도메인 개념
 * - 어떤 MFA 방식을 요구할지 정의
 * - 이미 완료된 MFA 추적
 */
data class MfaRequirement(
        val isRequired: Boolean,
        val requiredMethods: Set<MfaType>,
        val completedMethods: Set<MfaType>,
        val sessionId: String
) {
    /** 모든 필수 MFA가 완료되었는가? */
    fun isComplete(): Boolean {
        return completedMethods.containsAll(requiredMethods)
    }

    /** 남은 MFA 방식 */
    fun getRemainingMethods(): Set<MfaType> {
        return requiredMethods - completedMethods
    }

    companion object {
        fun none(sessionId: String): MfaRequirement {
            return MfaRequirement(false, emptySet(), emptySet(), sessionId)
        }

        fun otp(sessionId: String): MfaRequirement {
            return MfaRequirement(true, setOf(MfaType.OTP), emptySet(), sessionId)
        }

        fun withCompleted(
                sessionId: String,
                required: Set<MfaType>,
                completed: Set<MfaType>
        ): MfaRequirement {
            return MfaRequirement(true, required, completed, sessionId)
        }
    }
}
